client-runtime/src/main/java/com/microsoft/rest/v2/RestProxy.java

```diff
@@ -109,6 +112,28 @@ else if (returnTypeToken.isSubtypeOf(Single.class)) {
         }
         else {
             final HttpResponse response = httpClient.sendRequest(request);
+
+            final int responseStatusCode = response.statusCode();
+            if (!methodParser.isExpectedResponseStatusCode(responseStatusCode)) {
+                final Class<? extends RestException> exceptionType = methodParser.exceptionType();
+                try {
+                    final Class<?> exceptionBodyType = methodParser.exceptionBodyType();
+                    final Constructor<? extends RestException> exceptionConstructor = exceptionType.getConstructor(String.class, HttpResponse.class, exceptionBodyType);
+
+                    String responseContent = null;
+                    try {
+                        responseContent = response.bodyAsString();
+                    } catch (IOException ignored) {
+                    }
+
+                    final Object exceptionBody = responseContent == null ? null : serializer.deserialize(responseContent, exceptionBodyType);
+
+                    throw exceptionConstructor.newInstance("Status code " + responseStatusCode + ", " + responseContent, response, exceptionBody);
+                } catch (IllegalAccessException | InstantiationException | InvocationTargetException | NoSuchMethodException e) {
+                    throw new IOException("Status code " + responseStatusCode + ", but an instance of " + exceptionType.getCanonicalName() + " cannot be created.", e);
+                }
+            }
+
```

Comment:
The http status code `204` should not contain a body, so either we may want to skip deserialization in 204 case or see the behavior of `serializer.deserialize(null /** responseBodyString** /, returnType)`, if it throws on null then it's a problem. what do you think?

Reply:
On the line where I deserialize, I check if responseContent is null and only deserialize if it isn't null. I just added a check for if it is empty as well. I agree that we should be careful of NullPointerExceptions, but I don't think we'll have that problem here.